File: SMSPD_waveform_plot_ChatGPT.m
tic
% 資料夾路徑
folder_path ='E:\SNSPD\SNSPD_data\SMSPD_NbTiN_2025Apr\Laser\1-10\20250503\4.68\Pulse\800\80000kHz\0nW\0degrees\20250503_015313\Pulse_800_0nW_0degrees';

% 實驗參數
% Exp_para ='20241119_BroadbandSMSPD_P1_Pulse_450_8000nW_0degrees_';
% Exp_para ='SMSPD_NbTiN_1_1-1_Pulse_450_30000nW_0degrees_';

% 電壓範圍
% Va = [500:500:5500, 5600:100:7500] / 1000; % 將電壓從 mV 轉換為 V
% Va = [5,10:20,25,30]

file_list = dir(fullfile(folder_path, '*_mV.txt'));
if (isempty(file_list))
    error('No text file is found!')
end

%%
Va = sort(extract_mV_from_list(file_list));
[basename, ~] = extract_info(file_list(1).name);
Exp_para = basename;
% 可調整的總事件數
user_defined_event = 10001; % 使用者自定義的總事件數

% 優化的預分配記憶體儲存
DATA_LENGTH = 125.;
num_va = length(Va);
%max_s1_data = zeros(1000, num_va); % 每個電壓的最大 s1 數據
max_s1_data = zeros(DATA_LENGTH, num_va); % 每個電壓的最大 s1 數據
peak_max_values = zeros(1, num_va); % 儲存每個電壓的最大 peak 值

disp('Loading Data...')

% num_va = 4;

%% Loading Data
parfor k = 1:num_va
    % 加載數據
    % file_path = fullfile(folder_path, [Exp_para, num2str(Va(k)*1000), '_mV.txt']);
    % file_path = fullfile(folder_path, [Exp_para, num2str(Va(k)), '_mV.txt']);
    file_path = fullfile(folder_path,file_list(k).name)
    if ~isfile(file_path)
        warning('File not found: %s', file_path);
        continue;
    end
    signal = load(file_path, '-ascii'); % 快速加載數據

    % 確定事件數
    total_event = min(user_defined_event, floor(length(signal) / DATA_LENGTH)); % 根據數據長度計算事件數

    % 使用矩陣重整，避免逐事件迴圈
    reshaped_signal = reshape(signal(1:total_event*DATA_LENGTH), DATA_LENGTH, total_event); % 重整為 200 行的矩陣

    % 找到每個電壓的最大事件
    [max_val, max_idx] = max(max(reshaped_signal, [], 1)); % 獲取最大值和索引
    max_s1_data(:, k) = reshaped_signal(:, max_idx); % 儲存對應事件的完整數據

    % 計算最大 peak 值
    peak_max_values(k) = max_val;
end




%% plot 
disp('Generating plots...')
% 繪製所有電壓的最大 s1 數據圖
figure;
for k = 1:num_va
    plot(max_s1_data(:, k), 'DisplayName', ['Voltage ', num2str(Va(k)), ' mV']);
    hold on;
end
title('Max s1 Data Across Voltages');
xlabel('Data Index');
ylabel('s1 Value');
legend;
grid on;
savefig(fullfile(folder_path,[basename,'_waveform.fig']));
%savefig('waveform.fig')

% 繪製峰值對電壓的圖
figure;
plot(Va, peak_max_values, '-o');
title('Peak Max Value vs Voltage');
xlabel('Voltage (mV)');
ylabel('Peak Max Value');
grid on;
savefig(fullfile(folder_path,[basename,'_peakToVoltage.fig']));
%savefig('peak_voltage.fig')

%% save to txt file
disp('save data into txt file...')
F = [Va, peak_max_values.'];
save(fullfile(folder_path,[basename,'_peakToVoltage.txt']),'F','-ascii')

%%% End of the code
%% function block

function [basename, mV_value] = extract_info(filename)
    % Extracts the mV value from a filename string.
    % Example: 'SMSPD_NbTiN_1_1-1_Pulse_450_30000nW_0degrees_100_mV.txt'
    % Output: 100

    % Regular expression to find the mV value before '_mV'
    pattern = '(.*)_(\d+)_mV.txt';

    % Apply regular expression
    tokens = regexp(filename, pattern, 'tokens');

    % Check if a match is found
    if ~isempty(tokens)
        basename = tokens{1}{1};
        mV_value = str2double(tokens{1}{2}); % Convert extracted string to number
    else
        error(['No mV value found in filename. filename: ', filename]);
    end
end

function mV_values = extract_mV_from_list(filenames)
    % Extracts mV values from a cell array of filenames.
    % Input: cell array of filenames
    % Output: array of extracted mV values

    mV_values = NaN(size(filenames)); % Preallocate array for mV values

    parfor i = 1:length(filenames)
        [~, mV_values(i)] = extract_info(filenames(i).name);
        %% mV_values(i) = extract_mV(filenames(i));
    end
end

function y = triple(x)
    y = 3*x;
end
toc

File: SMSPD_waveform_plot_ChatGPT.py
"""
波形繪圖（對應 SMSPD_waveform_plot_ChatGPT.m）

加速版：用多核心平行讀取與計算各電壓的最大波形，主程式統一繪圖。

執行方式：
    1) 不帶參數 — 使用「設定區」的 folder_path
       py -3.8 SMSPD_waveform_plot_ChatGPT.py

    2) 指定資料夾
       py -3.8 SMSPD_waveform_plot_ChatGPT.py -d "C:\\path\\to\\folder"

    3) 指定單一 .txt 檔（只畫該檔的波形，不畫 peak vs voltage）
       py -3.8 SMSPD_waveform_plot_ChatGPT.py -i "C:\\path\\to\\file_mV.txt"
"""

import os
import re
import time
import argparse
import numpy as np
import matplotlib.pyplot as plt
from multiprocessing import Pool, cpu_count


# =====================================================================
# 設定區 — 請依實驗環境修改
# =====================================================================

folder_path = r'E:\SMSPD_NbTiN_1\Laser\1-1\20250107\4p5\Pulse\450\10000kHz\5000nW\0degrees\20250107_230159\Pulse_450_5000nW_0degrees'

# 每個電壓檔最多看幾個事件
user_defined_event = 10001

# 一個事件的取樣點數
DATA_LENGTH = 1000

# 平行處理用幾個核心。-1 表示用滿全部
NUM_WORKERS = -1


# =====================================================================
# 工具函式
# =====================================================================
def extract_info(filename):
    """從檔名抽出 basename 和 mV 值"""
    match = re.search(r'(.*)_(\d+)_mV\.txt', filename)
    if not match:
        raise ValueError(f'檔名格式不符: {filename}')
    return match.group(1), int(match.group(2))


# =====================================================================
# 工作函式：讀一個 txt，回傳「最大波形」與「峰值」
# =====================================================================
def process_one_file(args):
    file_path, voltage, user_defined_event, DATA_LENGTH = args

    raw = np.loadtxt(file_path)
    signal = raw[:, 0] if raw.ndim == 2 else raw

    # 計算實際事件數（不超過使用者上限，且不超過資料長度）
    total_event = min(user_defined_event, len(signal) // DATA_LENGTH)

    # 重塑成 (DATA_LENGTH, total_event) 的 2D 陣列（column-major，跟 Matlab 一致）
    reshaped = np.reshape(
        signal[:total_event * DATA_LENGTH],
        (DATA_LENGTH, total_event),
        order='F',
    )

    # 找出哪一個事件的峰值最大
    column_max = np.max(reshaped, axis=0)
    max_idx = int(np.argmax(column_max))
    max_val = column_max[max_idx]
    max_waveform = reshaped[:, max_idx]

    return voltage, max_waveform, max_val


# =====================================================================
# 主程式
# =====================================================================
if __name__ == '__main__':
    t_start = time.time()

    # --- CLI 覆寫 ---
    parser = argparse.ArgumentParser(description='SMSPD waveform plot')
    group = parser.add_mutually_exclusive_group()
    group.add_argument('-i', '--input', metavar='FILE', help='單一 *_mV.txt 檔')
    group.add_argument('-d', '--dir',   metavar='DIR',  help='含 *_mV.txt 的資料夾')
    cli = parser.parse_args()

    if cli.input:
        folder_path = os.path.dirname(os.path.abspath(cli.input))
        all_files = [os.path.basename(cli.input)]
    elif cli.dir:
        folder_path = cli.dir
        all_files = [f for f in os.listdir(folder_path) if f.endswith('_mV.txt')]
    else:
        all_files = [f for f in os.listdir(folder_path) if f.endswith('_mV.txt')]

    if not all_files:
        raise RuntimeError('No text file is found!')

    file_with_voltage = [(f, extract_info(f)[1]) for f in all_files]
    file_with_voltage.sort(key=lambda x: x[1])
    Va = np.array([v for _, v in file_with_voltage])
    basename = extract_info(file_with_voltage[0][0])[0]
    num_va = len(Va)

    # 決定 worker 數
    n_workers = cpu_count() if NUM_WORKERS == -1 else NUM_WORKERS
    print(f'使用 {n_workers} 個核心平行處理 {num_va} 個檔案')

    # 平行處理：每個 worker 讀一個檔、回傳該電壓的最大波形
    args_list = [
        (os.path.join(folder_path, f), v, user_defined_event, DATA_LENGTH)
        for f, v in file_with_voltage
    ]
    with Pool(processes=n_workers) as pool:
        results = pool.map(process_one_file, args_list)

    # 整理結果到固定大小的陣列
    max_s1_data = np.zeros((DATA_LENGTH, num_va))
    peak_max_values = np.zeros(num_va)
    for k, (_, waveform, max_val) in enumerate(results):
        max_s1_data[:, k] = waveform
        peak_max_values[k] = max_val

    # =================================================================
    # 繪圖：所有電壓的最大波形
    # =================================================================
    print('Generating plots...')
    plt.figure(figsize=(10, 6))
    for k in range(num_va):
        plt.plot(max_s1_data[:, k], label=f'{Va[k]} mV')
    plt.title('Max s1 Data Across Voltages')
    plt.xlabel('Data Index')
    plt.ylabel('s1 Value')
    plt.grid(True)
    plt.legend(fontsize=8, loc='best')
    waveform_png = os.path.join(folder_path, f'{basename}_waveform.png')
    plt.savefig(waveform_png, dpi=150, bbox_inches='tight')
    plt.close()
    print(f'  存檔: {waveform_png}')

    # =================================================================
    # 繪圖：峰值對電壓
    # =================================================================
    plt.figure(figsize=(8, 5))
    plt.plot(Va, peak_max_values, '-o')
    plt.title('Peak Max Value vs Voltage')
    plt.xlabel('Voltage (mV)')
    plt.ylabel('Peak Max Value')
    plt.grid(True)
    peakV_png = os.path.join(folder_path, f'{basename}_peakToVoltage.png')
    plt.savefig(peakV_png, dpi=150, bbox_inches='tight')
    plt.close()
    print(f'  存檔: {peakV_png}')

    # =================================================================
    # 輸出 peak vs voltage 數值資料
    # =================================================================
    F = np.column_stack([Va, peak_max_values])
    peakV_txt = os.path.join(folder_path, f'{basename}_peakToVoltage.txt')
    np.savetxt(peakV_txt, F, fmt='%14.7e')
    print(f'  存檔: {peakV_txt}')

    print(f'Done. Elapsed: {time.time() - t_start:.2f} s')
